runtime/io.h

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Defines API between compiled code and I/O runtime library.
+
+#ifndef FORTRAN_RUNTIME_IO_H_
+#define FORTRAN_RUNTIME_IO_H_
+
+#include "entry-names.h"
+#include <cinttypes>
+#include <cstddef>
+
+namespace Fortran::runtime {
+class Descriptor;
+class NamelistGroup;
+};
+
+namespace Fortran::runtime::IO {
+
+class IOStatementState;
+using Cookie = IOStatementState *;
+using ExternalUnit = int;
+using AsynchronousId = int;
+static constexpr ExternalUnit DefaultUnit{-1};  // READ(*), WRITE(*), PRINT
+
+extern "C" {
+
+#define IONAME(name) RTNAME(IO_##name)
+
+// These functions initiate data transfer statements (READ, WRITE, PRINT).
+// Example: PRINT *, 666 is implemented as the series of calls:
+//   Cookie cookie{BeginExternalListOutput(DefaultUnit)};
+//   OutputInteger64(cookie, 666);
```

Comment:
Why is "OutputInteger64(cookie, 666)" using 64 instead of say 32?  Why not use the kind & expect the runtime to sync with the compiler?  How will float16 and bfloat be differentiated?

Reply:
It's always safe to cast shorter integers up to 64 bits when passing their values as C arguments for output.

In f18, 16-bit IEEE floating-point is kind=2 and truncated 32-bit floating-point is kind=3, and these are distinguished in the descriptor.

Reply:
Assume that we don't care about targets that require two conversion to go from 16- to 64-bits (if such a target exists).  What's the plan when 128-bit integers come along?  What's the goal of just having the 64-bit entry point?

Reply:
*Any* supported type can be passed with a descriptor, including 16-bit and 8-bit and 128-bit integers.  The intent of the convenience routine for emitting a single 64-bit integer scalar is to avoid having to use a descriptor for most integer scalars.

Reply:
got it.  cool.